src/compiler/checker.ts

```diff
@@ -6785,30 +6830,175 @@ module ts {
                     if (constraint) {
                         typeArgumentsAreAssignable = checkTypeAssignableTo(typeArgument, constraint, reportErrors ? typeArgNode : undefined,
-                            Diagnostics.Type_0_does_not_satisfy_the_constraint_1);
+                            Diagnostics.Type_0_does_not_satisfy_the_constraint_1, containingMessageChain);
                     }
                 }
             }
             return typeArgumentsAreAssignable;
         }
+        
+        function getTypeOfParentOfClassElement(node: ClassElement) {
+            let classSymbol = getSymbolOfNode(node.parent);
+            if (node.flags & NodeFlags.Static) {
+                return getTypeOfSymbol(classSymbol);
+            }
+            else {
+                return getDeclaredTypeOfSymbol(classSymbol);
+            }
+        }
+        
+        function createTypedPropertyDescriptorType(propertyType: Type): Type {
+            let globalTypedPropertyDescriptorType = getGlobalTypedPropertyDescriptorType();
+            return globalTypedPropertyDescriptorType !== emptyObjectType 
+                ? createTypeReference(<GenericType>globalTypedPropertyDescriptorType, [propertyType]) 
+                : emptyObjectType;
+        }
+        
+        function getDecoratorTargetArgumentType(target: Node): Type {
+            // The first argument to a decorator is its `target`.
+            switch (target.kind) {
+                case SyntaxKind.ClassDeclaration:
+                    // For a class decorator, the `target` is the type of the class (e.g. the 
+                    // "static" or "constructor" side of the class)
+                    let classSymbol = getSymbolOfNode(target);
+                    return getTypeOfSymbol(classSymbol);
+                
+                case SyntaxKind.Parameter:
+                    // For a parameter decorator, the `target` is the parent type of the 
+                    // parameter's containing method. 
+                    target = target.parent;
+                    if (target.kind === SyntaxKind.Constructor) {
+                        let classSymbol = getSymbolOfNode(target);
+                        return getTypeOfSymbol(classSymbol);
+                    }
+                    
+                    // fall-through
+                    
+                case SyntaxKind.PropertyDeclaration:
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.GetAccessor:
+                case SyntaxKind.SetAccessor:
+                    // For a property or method decorator, the `target` is the
+                    // "static"-side type of the parent of the member if the member is
+                    // declared "static"; otherwise, it is the "instance"-side type of the 
+                    // parent of the member.
+                    return getTypeOfParentOfClassElement(<ClassElement>target);

```

Comment:
But these could also be on an object literal, at least in the case of accessors. Maybe you should bail out if it's an object literal before assuming it's a class.